src/lib/MAVLink/ardupilot_custom_telemetry.cpp

```diff
@@ -0,0 +1,330 @@
+#include "ardupilot_custom_telemetry.h"
+#include "common/mavlink.h"
+
+/*
+ * Known Issues:
+ * - Battery Capacity is currently not available via Mavlink BATTERY_INFO message is not implemented on Ardupilot side.
+ *   This can be worked around by setting the battery capacity via Yaapu Config option.
+ * - Distance to Home and Bearing to Home are highly volatile until armed due to updates
+ *   of both HOME Position and Current Position with low GPS accuracy. Ardupilot stops sending updates of HOME Position when armed.
+ */
+
+#define MIN(a,b) ((a)<(b)?(a):(b))
+#define MAX(a,b) ((a)>(b)?(a):(b))
+#define constrain(amt,low,high) ((amt)<(low)?(low):((amt)>(high)?(high):(amt)))
+
+#define NUMBER_OF_CORDIC_ITERATIONS (6U)
+/**
+ * static lookup for the angle corresponding tan(2^-i) * 1024
+ * essentially the degree that corresponds to a tan(degree) = [1, 0.5, 0.25, 0.125 0.0625, ...]
+ * scaled by 1024 to swap a (x/1000) with (x >> 10)
+ */
+static const uint16_t lookup_theta[NUMBER_OF_CORDIC_ITERATIONS] = {46080, 27203, 14373, 7296, 3662, 1832};
+
+/**
+ * COS(degree) of phi [0.7071067811865476, 0.8944271909999159, 0.9701425001453319, 0.9922778767136676, 0.9980525784828885, 0.9995120760870788]
+ * product of all factors 0.607351770141296
+ */
+
+/**
+ * Cordic Algo to transform cartesian coordinates into polar coordinates
+ * Takes dlat and dlon as int32_t in degE7 scaling.
+ * Writes corrected vector length and angle to output pointers.
+ * We aim for meter accuracy and degree accuracy.
+ */
+void cartesian_to_polar_coordinates(int32_t dlat, int32_t dlon, int32_t *out_phi_deg, int32_t *out_radius_m)
+{
+    int8_t sign = 0;
+    int32_t theta = 0;
+    int32_t x = dlon;
+    int32_t y = dlat;
+    int32_t nx = 0;
+    int32_t ny = 0;    
+    int32_t phi = 0;
+    int32_t radius = 0;
+
+    //quadrant analysis
+    uint8_t quadrant = 1;
+    if (x < 0){
+        x = -x;
+        if (y < 0){
+            y = -y;
+            quadrant = 3;
+        } else {
+            quadrant = 2;
+        }
+    } else if (y < 0){
+        y = -y;
+        quadrant = 4;
+    }
+    
+    for (uint8_t index=0; index < NUMBER_OF_CORDIC_ITERATIONS; index++){
+        sign = (y >= 0) ? 1 : -1;
+        nx = x + (y >> index) * sign;
+        ny = y - (x >> index) * sign;
+        theta += lookup_theta[index] * sign;
+        x = nx;
+        y = ny;
+    }
+
+    // quadrant based angle correction
+    if (quadrant == 1){
+        phi = 90 - (theta >> 10);
+    } else if (quadrant == 2){
+        phi = 270 + (theta >> 10);
+    } else if (quadrant == 3){
+        phi = 270 - (theta >> 10);
+    } else {
+        phi = 90 + (theta >> 10);
+    }
+    // Vector Length Correction and Latitude/Longitude to Meter Conversion 
+    // Factor for COS Correction 0.607351770141296
+    // Factor for LatLonDegE7 to M Conversion 0.011131884502145034
+    // Combined 0.00676097 ~= 443/65536
+    radius = (x * 443) >> 16;
```

Comment:
I have to check this calc again and verify by hand.

Heading is within expected accuracy but distance is too high by 10-50%.

Reply:
ok, i know what it is. I forgot to scale longitude. Have to check how I do that with cordic.
```
ftype longitude_scale(int32_t lat);
ftype longitude_scale(int32_t lat)
{
    ftype scale = cos(lat * (1.0e-7 * DEG_TO_RAD));
    return MAX(scale, 0.01);
}
```

Reply:
Done, was quite complicated and error prone in int32_t with degE7 scaling. Currently testing.